### PR TITLE
Outages Update

### PR DESCRIPTION
**Pull request by jpie**
Big PR!
- Adds group outage overview
- Adds statistics for groups and individual sites
- Adds error handling to db queries
- Many minor css updates to home page (spacing, colors, etc.)
- Changes to 'create a group' page

### CODE REVIEW

File: controllers/groups.js [hunk 1]
```diff
@@ -20,258 +18,222 @@ function groupsDefault(req, res5) {
...
-      }else{
+  return new Promise((resolve, reject) => {
+    try{
+      if(req.body.join){
+        db.query(`INSERT INTO group_users(group_id, user_id)
+        SELECT group_id, user_id
+        FROM groups, users
+        WHERE groups.groupname = '${groupname}' and users.user_id = '${user_id}'`,
+        function(err){
+          if(err) return reject(err);
```
Comment (trycrmr) [inline controllers/groups.js]: return reject(error) does not appear to properly return an error to the browser, so try something else.

File: controllers/groups.js [hunk 2]
```diff
@@ -20,258 +18,222 @@ function groupsDefault(req, res5) {
...
-        if(host[host.length-1] == '/')    
-          host = host.slice(0, -1)
+      if(req.body.leave){
+        db.query(`delete from group_users
+        where user_id = '${user_id}' and group_id = (
+        select group_id
+        from groups
+        where groupname = '${groupname}')`,
+        function(err){
+          if (err) return reject(err);
```
Comment (trycrmr) [inline controllers/groups.js]: return reject(error) does not appear to properly return an error to the browser, so try something else.

File: controllers/groups.js [hunk 3]
```diff
@@ -20,258 +18,222 @@ function groupsDefault(req, res5) {
...
-                FROM endpoints WHERE url = '${host}')`,
-                function(error, results){
-                  if(error) console.log(error);
-        //Inserts url into database
-          db.query("SELECT endpoint_id FROM endpoints WHERE url = '"+host+"'", function(err, res){
+      db.query(`SELECT group_id
+      FROM groups
+      WHERE groupname = '${groupname}'`,
+      function(err, res3){
+        if(err) return reject(err);
```
Comment (trycrmr) [inline controllers/groups.js]: return reject(error) does not appear to properly return an error to the browser, so try something else.

File: controllers/groups.js [hunk 4]
```diff
@@ -20,258 +18,222 @@ function groupsDefault(req, res5) {
...
+          WHERE endpoint_id IN 
+          (SELECT endpoint_id 
+          FROM endpoints 
+          WHERE group_id IN 
+          (SELECT group_id 
+          FROM groups 
+          WHERE endpoints.url = '${hostDelete}' 
+          AND groups.group_id = '${group_id}'))`,
+              function(error, results){
+                if(error) return reject(error);
```
Comment (trycrmr) [inline controllers/groups.js]: return reject(error) does not appear to properly return an error to the browser, so try something else.

File: controllers/groups.js [hunk 5]
```diff
@@ -20,258 +18,222 @@ function groupsDefault(req, res5) {
...
+            host = host.replace(x, "")
+          if(host[host.length-1] == '/')    
+            host = host.slice(0, -1)
+
+          db.query(`INSERT INTO endpoints(url, currStatus)
+                  SELECT '${host}', 0
+                  WHERE NOT EXISTS (SELECT url 
+                  FROM endpoints WHERE url = '${host}')`,
+                  function(error, results){
+                    if(error) return reject(error);
```
Comment (trycrmr) [inline controllers/groups.js]: return reject(error) does not appear to properly return an error to the browser, so try something else.

File: controllers/groups.js [hunk 6]
```diff
@@ -20,258 +18,222 @@ function groupsDefault(req, res5) {
...
+          //Inserts url into database
+            db.query("SELECT endpoint_id FROM endpoints WHERE url = '"+host+"'", function(err, res){
+
+              var e = JSON.stringify(res.rows[0].endpoint_id);
+              console.log(e)
```
Comment (trycrmr) [inline controllers/groups.js]: Avoid leaving extra console.logs in PRs.

File: controllers/groups.js [hunk 7]
```diff
@@ -20,258 +18,222 @@ function groupsDefault(req, res5) {
...
+            //Defines endpoint Id
+              db.query(`INSERT INTO end_groups(endpoint_id, group_id, failedPings) 
+              SELECT '${e}', '${group_id}', '${failedPings}' 
+              WHERE not exists 
+              (SELECT endpoint_id, user_id 
+              FROM end_users 
+              WHERE endpoint_id = '${e}' 
+              AND user_id = '${group_id}')`,
+              function(err, res){
+                if(err) return reject(err);
```
Comment (trycrmr) [inline controllers/groups.js]: return reject(error) does not appear to properly return an error to the browser, so try something else.

File: controllers/groups.js [hunk 8]
```diff
@@ -20,258 +18,222 @@ function groupsDefault(req, res5) {
...
+              //Tracks sites for users
```
Comment (trycrmr) [inline controllers/groups.js]: I like this! Add more comments describing your code in the future.

File: controllers/groups.js [hunk 9]
```diff
@@ -20,258 +18,222 @@ function groupsDefault(req, res5) {
...
-    
-    })
-  res.redirect('/group/'+groupname);
+        }
+      
+      })
+      res.redirect('/group/'+groupname);
+    }catch(e){
+      console.log(e)
+      return reject(e)
```
Comment (trycrmr) [inline controllers/groups.js]: return reject(error) does not appear to properly return an error to the browser, so try something else.

File: controllers/home.js [hunk 10]
```diff
@@ -1,114 +1,120 @@
 var db = require('../db');
 
 function homeDefault(req1, res1) {
-  db.query("SELECT sess FROM sessions", function(err, res){
+  return new Promise((resolve, reject) => {
+    try {
+      if(req1.user == undefined){
+        console.log('Not logged in')
+        res1.render('auth/login', { title: 'Login or Register'})
```
Comment (trycrmr) [inline controllers/home.js]: Is there any message presented to the user here?

File: controllers/home.js [hunk 11]
```diff
@@ -1,114 +1,120 @@
...
+          left outer join groups on
+          (end_groups.group_id = groups.group_id)
+          where groups.group_id in (
+          select group_id
+          from group_users
+          where user_id = '${user_id}')
+          group by groupname
+          order by groupname`,
+          function(err, res2){
+            if (err) return reject(err)
```
Comment (trycrmr) [inline controllers/home.js]: return reject(error) does not appear to properly return an error to the browser, so try something else.

File: controllers/home.js [hunk 12]
```diff
@@ -1,114 +1,120 @@
...
-                return cvs;
+              db.query(`SELECT groupname
+              FROM groups
+              WHERE group_id in
+              (SELECT group_id 
+              FROM group_users
+              WHERE group_users.user_id = '${user_id}')
+              ORDER BY groupname ASC`,
+              function(err, res3){
+                if (err) return reject(err)
```
Comment (trycrmr) [inline controllers/home.js]: return reject(error) does not appear to properly return an error to the browser, so try something else.

File: controllers/home.js [hunk 13]
```diff
@@ -1,114 +1,120 @@
...
+                  where public = true
+                  AND
+                  group_id not in(
+                  select group_id 
+                  from group_users
+                  where user_id = '${user_id}')
+                  ORDER BY groups.createdAt DESC
+                  LIMIT 5`,
+                  function(err, res4){
+                    if (err) return reject(err)
```
Comment (trycrmr) [inline controllers/home.js]: return reject(error) does not appear to properly return an error to the browser, so try something else.

File: controllers/data.js [hunk 14]
```diff
@@ -1,58 +1,133 @@
 var db = require('../db');
 var url = require('url');
```
Comment (goterpsgo) [inline controllers/data.js]: It'd be helpful to have some 10000ft level comments to note what your code blocks are doing, esp where it relates to how any initial results are handled later on in any given method.

File: controllers/data.js [hunk 15]
```diff
@@ -1,58 +1,133 @@
...
 
-function dataDefault(req, res1, next) {
-  if(req.user == undefined){
+function dataUser(req, res1, next) {
```
Comment (goterpsgo) [inline controllers/data.js]: It'd be helpful to have some 10000ft level comments to note what your code blocks are doing, esp where it relates to how any initial results are handled later on in any given method.

File: controllers/groups.js [hunk 16]
```diff
@@ -20,258 +18,222 @@ function groupsDefault(req, res5) {
...
+        myEmail = res2.rows[0].email
+        if (err) return reject(err);
+
+        db.query(`SELECT creator
+        FROM groups
+        WHERE groupname = '${groupname}'`,
+        function(err, res3){
+          if (err) return reject(err);
+          if(res3.rows[0] == undefined){
+            res5.render('errors/grouperr')
```
[the remark this answers is not in the document]
Comment (goterpsgo) [reply controllers/groups.js]: Looks like results from the last one (res5) is what is used to return to the calling method/event. It took a bit of tracing and and the logic is a bit hard to trace through.

File: controllers/groups.js [hunk 17]
```diff
@@ -20,258 +18,222 @@ function groupsDefault(req, res5) {
...
-                    });
-                  }
-                })
-              })        
+  return new Promise((resolve, reject) => {
+    try {
+      db.query(`SELECT email
+      FROM users
+      WHERE user_id = '${user_id}'`,
+      function(err, res2){
```
Comment (goterpsgo) [inline controllers/groups.js]: I noticed the callback functions are nested five deep. We should consider breaking out unnamed inline functions and make each their own named methods (unless I'm reading this wrong).

File: pingserver.js [hunk 18]
```diff
@@ -10,315 +10,347 @@ dotenv.config();
...
-              FROM endpoints
-              WHERE endpoints.url = '${endpoints[r]}' AND end_groups.failedPings <= '${badPing}'))))`,
-              function(req, res){
-                var notify = res.rows.reduce(function(cvs, thisData){
-                  cvs[0].x.push(thisData.email);
-                  return cvs;
-
-                }, [{x: []}])
-                var k = JSON.stringify(notify[0].x)
+              db.query(`INSERT INTO uptime(endpoint_id, latency, status, date)
```
Comment (goterpsgo) [inline pingserver.js]: This action could be a candidate for being refactored and encapsulated into its own method since it's used more than once.

**Comment by trycrmr** [comment]
Closing as these changes have already been merged into dev.

**Pull request closed without merging**